### DISCUSSION TITLE
MultiExecutor on 2 isolates separated from main

### DISCUSSION THREAD

**Discussion by davidmartos96** [simolus3/drift#2748] [Q&A]
Hello!
I'm trying to avoid locks on readers of the database while performing sync writes that come from the backend.

In order to do that I was thinking about enabling WAL mode and then opening 2 connections to the database (read+write) with MultiExecutor. Since I haven't found any examples in the docs I was wondering if this would be correct.

```dart
DatabaseConnection connect() {
  return DatabaseConnection.delayed(Future(() async {
    final dbPath = await getDatabasePath();
    print("Using todos database at path $dbPath");

    QueryExecutor dbOpener() {
      return NativeDatabase(File(dbPath));
    }

    final isolateWrite = await DriftIsolate.spawn(
      dbOpener,
      isolateSpawn: getSpawnerWithName("dbWriter"),
    );
    final backgroundWriteConn = await isolateWrite.connect();

    final isolateRead = await DriftIsolate.spawn(
      dbOpener,
      isolateSpawn: getSpawnerWithName("dbReader"),
    );
    final backgroundReadConn = await isolateRead.connect();

    final multiexec = MultiExecutor(
      read: backgroundReadConn.executor,
      write: backgroundWriteConn.executor,
    );

    return backgroundWriteConn.withExecutor(multiexec);
  }));
}
```

I have some doubts about the last line, "withExecutor". Does it matter if I use writeConn vs readConn? Or is there a better way to create an effective connection with the `MultiExecutor` under the hood?

Additionally, where would you recommend calling `PRAGMA journal_mode=WAL;`? In the `setup` or `beforeOpen`? I'm guessing `setup` so that all isolates are configured as such. But since WAL updates the database file maybe it doesn't make any difference.

For my usecase I need the `reader` in a separate isolate as there are some heavy queries which we would like to avoid blocking the UI thread. Unfortunately I haven't seen any examples using 2 separate isolates

**Comment by simolus3** [comment]
> I have some doubts about the last line, "withExecutor". Does it matter if I use writeConn vs readConn? Or is there a better way to create an effective connection with the `MultiExecutor` under the hood?

As long as you are consistent it should be fine - a connection in drift is the combination of the underlying executor and something to broadcast updates for stream queries. If you only intend to have one database instance against this connection, you can also pass `singleClientMode: true` to [connect()](https://drift.simonbinder.eu/api/isolate/driftisolate/connect) in which case stream queries aren't sent over the connection at all (since we assume there will be no other instance of the database object).

> Additionally, where would you recommend calling `PRAGMA journal_mode=WAL;`? In the `setup` or `beforeOpen`? I

I think doing it in `setup` is safer since it's the first thing that happens to an opened sqlite3 database.

> For my usecase I need the `reader` in a separate isolate as there are some heavy queries which we would like to avoid blocking the UI thread. Unfortunately I haven't seen any examples using 2 separate isolates

I think using a `MultiExecutor` with two connections backed by different isolates is a great match for that. You could try sharing the actual sqlite3 database connection between the two isolates (open the database with `package:sqlite3/`, obtain the pointer via `database.handle`, send it to the other isolate and do `sqlite3.fromPointer` in `dbOpener`). But I don't actually know whether that will improve performance.

**Reply by davidmartos96** [reply]
Thank you Simon!
What do you mean by consistency? When would I need to do this again to watch out for that? If the drift database uses the custom "connect" function it should be ok, no?

What's the use case for singleClientMode = false? Doesn't drift warn when a second instance of a drift database is created?

Thanks for the idea on passing the pointer to the isolates. I'm guessing the difference if we were talking about native and not dart, would be using sqlite multithreaded in a single process vs multiprocess with 1 connection per process. Is that correct?
I would have to test if there are performance differences.

**Reply by davidmartos96** [reply]
@simolus3 Is this what you were referring to with the pointer so that they share the connection?
This will be ideal actually, because we sometimes depend on TEMP tables and ATTACHed databases for our business logic, so having separate connections doesn't go well with that.

```dart
    final nativeDb = sqlite3.open(_file.path);

    final ptrAddress = nativeDb.handle.address;

    QueryExecutor dbOpener() {
      final isolateNativeDb =
          sqlite3.fromPointer(Pointer.fromAddress(ptrAddress));
      return NativeDatabase.opened(isolateNativeDb);
    }

    final isolateWrite = await DriftIsolate.spawn(
      dbOpener,
      isolateSpawn: getSpawnerWithName("dbWriter"),
    );
    final backgroundWriteConn = await isolateWrite.connect();

    final isolateRead = await DriftIsolate.spawn(
      dbOpener,
      isolateSpawn: getSpawnerWithName("dbReader"),
    );
    final backgroundReadConn = await isolateRead.connect();

    final multiexec = MultiExecutor(
      read: backgroundReadConn.executor,
      write: backgroundWriteConn.executor,
    );
    ....
```
Unfortunately I cannot get it to work. Passing the pointer doesn't seem to be supported, and passing the int address through the isolate fails with a hard crash.

Error dump with pointer:
```
Invalid argument(s): Illegal argument in isolate message: (object is a Pointer)
   <- field ptr in main.<anonymous closure>.dbOpener (from file:///home/david/Dev/Flutter/drift/drift/test/engines/wal_mode_test.dart)
   <- _List len:2 (from dart:core)
   <- Instance(length:2) of '_GrowableList' (from dart:core)
  
  dart:isolate                           Isolate.spawn
  test/engines/wal_mode_test.dart 53:24  main.<fn>.getSpawnerWithName.spawn
  package:drift/isolate.dart 163:23      DriftIsolate.spawn
  test/engines/wal_mode_test.dart 75:45  main.<fn>
```

Error dump with int address:
```
===== CRASH =====
si_signo=Segmentation fault(11), si_code=1, si_addr=(nil)
version=3.2.0 (stable) (Tue Nov 14 18:26:59 2023 +0000) on "linux_x64"
pid=173233, thread=173277, isolate_group=main(0x7f00e8000fe0), isolate=dbWriter(0x7f00e4000db0)
os=linux, arch=x64, comp=no, sim=no
isolate_instructions=557d3b1b3b40, vm_instructions=557d3b1b3b40
fp=7f00f1ffde60, sp=7f00f1ffde18, pc=0
  pc 0x0000000000000000 fp 0x00007f00f1ffde60 Unknown symbol
  pc 0x00007f00f1c6a5be fp 0x00007f00f1ffdea0 sqlite3_create_function_v2+0x1e
  pc 0x00007f0113186abb fp 0x00007f00f1ffdee0 Unknown symbol
  pc 0x00007f00f285c6a6 fp 0x00007f00f1ffdf50 Unknown symbol
  pc 0x00007f00f285c19f fp 0x00007f00f1ffe020 Unknown symbol
  pc 0x00007f00f285aa62 fp 0x00007f00f1ffe100 Unknown symbol
  pc 0x00007f00f2859ef9 fp 0x00007f00f1ffe1a8 Unknown symbol
  pc 0x00007f00f2859708 fp 0x00007f00f1ffe1f8 Unknown symbol
  pc 0x00007f00f285927d fp 0x00007f00f1ffe240 Unknown symbol
  pc 0x00007f00f2858c0a fp 0x00007f00f1ffe2a8 Unknown symbol
  pc 0x00007f00f2858639 fp 0x00007f00f1ffe2f8 Unknown symbol
  pc 0x00007f00f2858579 fp 0x00007f00f1ffe3a0 Unknown symbol
  pc 0x00007f00f2856d43 fp 0x00007f00f1ffe3e0 Unknown symbol
  pc 0x00007f00f2855e16 fp 0x00007f00f1ffe448 Unknown symbol
  pc 0x00007f00f2855ad9 fp 0x00007f00f1ffe478 Unknown symbol
  pc 0x00007f00f349f4f6 fp 0x00007f00f1ffe4e0 Unknown symbol
  pc 0x00007f00f349f804 fp 0x00007f00f1ffe528 Unknown symbol
  pc 0x00007f00f349f4f6 fp 0x00007f00f1ffe590 Unknown symbol
  pc 0x00007f00f349e5ac fp 0x00007f00f1ffe5e8 Unknown symbol
  pc 0x00007f0113182e46 fp 0x00007f00f1ffe660 Unknown symbol
  pc 0x0000557d3b2f1a93 fp 0x00007f00f1ffe6c0 dart::DartEntry::InvokeFunction+0x163
  pc 0x0000557d3b2f34d6 fp 0x00007f00f1ffe700 dart::DartLibraryCalls::HandleMessage+0x126
  pc 0x0000557d3b30fb1f fp 0x00007f00f1ffec90 dart::IsolateMessageHandler::HandleMessage+0x2bf
  pc 0x0000557d3b331d47 fp 0x00007f00f1ffed00 dart::MessageHandler::HandleMessages+0x127
  pc 0x0000557d3b332338 fp 0x00007f00f1ffed50 dart::MessageHandler::TaskCallback+0x1e8
  pc 0x0000557d3b42f35b fp 0x00007f00f1ffedd0 dart::ThreadPool::WorkerLoop+0x14b
  pc 0x0000557d3b42f5f8 fp 0x00007f00f1ffee00 dart::ThreadPool::Worker::Main+0x78
  pc 0x0000557d3b3b8bf6 fp 0x00007f00f1ffeec0 dart::ThreadStart+0xd6
  pc 0x00007f011396c947 fp 0x00007f00f1ffef70 /lib64/libc.so.6+0x8c947
-- End of DumpStackTrace
  pc 0x0000000000000000 fp 0x00007f00f1ffdee0 sp 0x0000000000000000 Cannot find code object
  pc 0x00007f00f285c6a6 fp 0x00007f00f1ffdf50 sp 0x00007f00f1ffdef0 [Optimized] FfiTrampoline__sqlite3_create_function_v2
  pc 0x00007f00f285c19f fp 0x00007f00f1ffe020 sp 0x00007f00f1ffdf60 [Unoptimized] Bindings.sqlite3_create_function_v2
  pc 0x00007f00f285aa62 fp 0x00007f00f1ffe100 sp 0x00007f00f1ffe030 [Unoptimized] FfiDatabase.sqlite3_create_function_v2
  pc 0x00007f00f2859ef9 fp 0x00007f00f1ffe1a8 sp 0x00007f00f1ffe110 [Unoptimized] DatabaseImplementation.createFunction
  pc 0x00007f00f2859708 fp 0x00007f00f1ffe1f8 sp 0x00007f00f1ffe1b8 [Unoptimized] EnableNativeFunctions|useNativeFunctions
  pc 0x00007f00f285927d fp 0x00007f00f1ffe240 sp 0x00007f00f1ffe208 [Unoptimized] Sqlite3Delegate._initializeDatabase@262164674
  pc 0x00007f00f2858c0a fp 0x00007f00f1ffe2a8 sp 0x00007f00f1ffe250 [Unoptimized] new Sqlite3Delegate.opened
  pc 0x00007f00f2858639 fp 0x00007f00f1ffe2f8 sp 0x00007f00f1ffe2b8 [Unoptimized] new _NativeDelegate@21476956.opened
  pc 0x00007f00f2858579 fp 0x00007f00f1ffe3a0 sp 0x00007f00f1ffe308 [Unoptimized] new NativeDatabase.opened
  pc 0x00007f00f2856d43 fp 0x00007f00f1ffe3e0 sp 0x00007f00f1ffe3b0 [Unoptimized] main.<anonymous closure>.dbOpener
  pc 0x00007f00f2855e16 fp 0x00007f00f1ffe448 sp 0x00007f00f1ffe3f0 [Unoptimized] _startDriftIsolate@20180803
  pc 0x00007f00f2855ad9 fp 0x00007f00f1ffe478 sp 0x00007f00f1ffe458 [Unoptimized] _startDriftIsolate@20180803
  pc 0x00007f00f349f4f6 fp 0x00007f00f1ffe4e0 sp 0x00007f00f1ffe488 [Unoptimized] _Closure@0150898.dyn:call
  pc 0x00007f00f349f804 fp 0x00007f00f1ffe528 sp 0x00007f00f1ffe4f0 [Unoptimized] _delayEntrypointInvocation@1026248.<anonymous closure>
  pc 0x00007f00f349f4f6 fp 0x00007f00f1ffe590 sp 0x00007f00f1ffe538 [Unoptimized] _Closure@0150898.dyn:call
  pc 0x00007f00f349e5ac fp 0x00007f00f1ffe5e8 sp 0x00007f00f1ffe5a0 [Unoptimized] _RawReceivePort@1026248._handleMessage@1026248
  pc 0x00007f0113182e46 fp 0x00007f00f1ffe660 sp 0x00007f00f1ffe5f8 [Stub] InvokeDartCode
```

Is sharing the db handle connection between isolates something that you have managed to do before?

**Reply by davidmartos96** [reply]
I have tried in two versions of Sqlite, one with THREADSAFE=1, and another one with THREADSAFE=2, but both fail the same way.

**Reply by simolus3** [reply]
Scratch the idea with the shared connection used in different isolates - it will break with user-defined functions since those can only be called from the isolate registering the function. But with a shared database connection, calling `createFunction` will override the function for all threads. I thought it might work but I haven't considered this. I don't think this is fixable.

>  What do you mean by consistency? When would I need to do this again to watch out for that? If the drift database uses the custom "connect" function it should be ok, no?

In your setup it's completely fine, yes. Only if another database (perhaps created in another isolate too) wants to re-use the existing isolates, it should remember to use the right one to synchronize streams. But since that's not the case in your code, just do this and use `singleClientMode`:

```dart
    final multiexec = MultiExecutor(
      read: backgroundReadConn.executor,
      write: backgroundWriteConn.executor,
    );

    return DatabaseConnection(multiexec);
```

> What's the use case for singleClientMode = false? Doesn't drift warn when a second instance of a drift database is created?

Without `singleClientMode`, drift will announce all changes it makes to the remote isolate which would forward it to other clients. This is useful if you want to store the connect port of the isolate in a name registry and perhaps have other background isolates connect to it. Writes made by those isolates would then invalidate streams in the UI. Since you're not using that, you can make it not announce (and listen for) updates. Another effect of `singleClientMode` is that the isolates will be shut down once you close the database, since you promise that no other database will connect to it. Without that, the spawned isolates would stay alive until you call `shutdown` on them.

> I'm guessing the difference if we were talking about native and not dart, would be using sqlite multithreaded in a single process vs multiprocess with 1 connection per process. Is that correct?

Pretty much, yes. My intuition was that multithreaded might be faster because it can use locks in userspace instead of having to go through the file system, but with WAL it's using shared memory for writes either way so perhaps it's not that bad. The overhead of a second connection is definitely not as big as the performance benefits you get from splitting reads and writes into different isolates.

I haven't used this setup before, so let me know how it goes and whether it improves things for you.
